src/lib.rs

```diff
@@ -1,14 +1,207 @@
-pub fn add(left: usize, right: usize) -> usize {
-    left + right
+use std::{cmp::Ordering, hash};
+
+pub use proc_macros::grouped_ordering;
+
+pub trait GroupedOrdering {
+    type Group: hash::Hash + Eq;
+
+    fn compare(&self, a: &Self::Group, b: &Self::Group) -> Ordering;
+}
+
+pub trait GroupedOrderable<TGroupedOrdering: GroupedOrdering> {
+    fn map_to_grouped_ordering(&self) -> TGroupedOrdering::Group;
+}
+
+pub trait VecExt {
+    type Item;
+
+    fn sort_by_grouped_ordering<TGroupedOrdering>(&mut self, grouped_ordering: &TGroupedOrdering)
+    where
+        TGroupedOrdering: GroupedOrdering,
+        Self::Item: GroupedOrderable<TGroupedOrdering>;
+}
+
+impl<T> VecExt for Vec<T> {
+    type Item = T;
+
+    fn sort_by_grouped_ordering<TGroupedOrdering>(&mut self, grouped_ordering: &TGroupedOrdering)
+    where
+        TGroupedOrdering: GroupedOrdering,
+        Self::Item: GroupedOrderable<TGroupedOrdering>,
+    {
+        self.sort_by(|a, b| {
+            grouped_ordering.compare(&a.map_to_grouped_ordering(), &b.map_to_grouped_ordering())
+        });
+    }
 }
 
 #[cfg(test)]
 mod tests {
+    use std::collections::HashMap;
+
+    use speculoos::prelude::*;
+
     use super::*;
+    use proc_macros::grouped_ordering_crate_internal as grouped_ordering;
+
+    #[test]
+    fn test_manual_impl_vec_sort() {
+        #[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+        enum GroupedOrderingFooGroup {
+            A,
+            B,
+            C,
+        }
+
+        struct GroupedOrderingFoo {
+            #[allow(dead_code)]
+            groups: [GroupedOrderingFooGroup; 3],
+            index_lookup: HashMap<GroupedOrderingFooGroup, usize>,
```

Comment:
Given that for the `Vec`-sorting use case we're effectively doing vec item -> grouped-ordering group -> "group index" lookups to accomplish the sorting, it seemed silly to only have eg this `[GroupType; 3]` array and have to do array-lookups rather than a faster key -> value lookup

So basically `index_lookup` is just an optimization and is redundant with `groups`

I initially was combining them into a single `IndexMap` but then decided maybe it was weird to force that dependency on users of the crate so separated into two fields